File: Integraupt-Backend/backend-horarios-laravel/app/Http/Controllers/Api/HorarioController.php
<?php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Services\HorarioService;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;

/**
 * Controlador API para Horarios.
 * Replica: com.horarios.controller.HorarioController
 *
 * Endpoints:
 * GET    /api/horarios                           → index()
 * GET    /api/horarios/disponibles               → disponibles()
 * GET    /api/horarios/ocupados                  → ocupados()
 * GET    /api/horarios/espacio/{espacioId}       → porEspacio()
 * GET    /api/horarios/espacio/{espacioId}/semanal → semanal()
 * GET    /api/horarios/dia/{diaSemana}           → porDia()
 * GET    /api/horarios/{id}                      → show()
 * POST   /api/horarios                           → store()
 * PUT    /api/horarios/{id}                      → update()
 * PATCH  /api/horarios/{id}/ocupacion            → actualizarOcupacion()
 * DELETE /api/horarios/{id}                      → destroy()
 */
class HorarioController extends Controller
{
    public function __construct(
        private readonly HorarioService $horarioService,
    ) {}

    /**
     * GET /api/horarios
     * Replica: @GetMapping → listarHorarios()
     */
    public function index(): JsonResponse
    {
        return response()->json($this->horarioService->listarTodos());
    }

    /**
     * GET /api/horarios/{id}
     * Replica: @GetMapping("/{id}") → obtenerHorario()
     */
    public function show(int $id): JsonResponse
    {
        $horario = $this->horarioService->buscarPorId($id);

        if (!$horario) {
            return response()->json([
                'exito' => false,
                'mensaje' => 'Horario no encontrado',
            ], 404);
        }

        return response()->json([
            'exito' => true,
            'mensaje' => 'Horario encontrado',
            'horario' => $horario,
        ]);
    }

    /**
     * GET /api/horarios/espacio/{espacioId}
     * Replica: @GetMapping("/espacio/{espacioId}") → listarPorEspacio()
     */
    public function porEspacio(int $espacioId): JsonResponse
    {
        return response()->json($this->horarioService->listarPorEspacio($espacioId));
    }

    /**
     * GET /api/horarios/espacio/{espacioId}/semanal
     * Replica: @GetMapping("/espacio/{espacioId}/semanal") → obtenerHorarioSemanal()
     */
    public function semanal(int $espacioId): JsonResponse
    {
        return response()->json($this->horarioService->obtenerHorarioSemanalPorEspacio($espacioId));
    }

    /**
     * GET /api/horarios/dia/{diaSemana}
     * Replica: @GetMapping("/dia/{diaSemana}") → listarPorDia()
     */
    public function porDia(string $diaSemana): JsonResponse
    {
        try {
            return response()->json($this->horarioService->listarPorDia($diaSemana));
        } catch (\InvalidArgumentException $e) {
            return response()->json([
                'exito' => false,
                'mensaje' => $e->getMessage(),
            ], 400);
        }
    }

    /**
     * GET /api/horarios/disponibles
     * Replica: @GetMapping("/disponibles") → listarDisponibles()
     */
    public function disponibles(): JsonResponse
    {
        return response()->json($this->horarioService->listarPorOcupacion(false));
    }

    /**
     * GET /api/horarios/ocupados
     * Replica: @GetMapping("/ocupados") → listarOcupados()
     */
    public function ocupados(): JsonResponse
    {
        return response()->json($this->horarioService->listarPorOcupacion(true));
    }

    /**
     * POST /api/horarios
     * Replica: @PostMapping → crearHorario()
     */
    public function store(Request $request): JsonResponse
    {
        $validated = $request->validate([
            'espacioId' => 'required|integer',
            'bloqueId' => 'required|integer',
            'diaSemana' => 'required|string',
            'ocupado' => 'nullable|boolean',
        ]);

        try {
            $horario = $this->horarioService->crearHorario($validated);
            return response()->json([
                'exito' => true,
                'mensaje' => 'Horario creado correctamente',
                'horario' => $horario,
            ], 201);
        } catch (\InvalidArgumentException $e) {
            return response()->json([
                'exito' => false,
                'mensaje' => $e->getMessage(),
            ], 400);
        }
    }

    /**
     * PUT /api/horarios/{id}
     * Replica: @PutMapping("/{id}") → actualizarHorario()
     */
    public function update(Request $request, int $id): JsonResponse
    {
        $validated = $request->validate([
            'espacioId' => 'required|integer',
            'bloqueId' => 'required|integer',
            'diaSemana' => 'required|string',
            'ocupado' => 'nullable|boolean',
        ]);

        try {
            $horario = $this->horarioService->actualizarHorario($id, $validated);
            return response()->json([
                'exito' => true,
                'mensaje' => 'Horario actualizado correctamente',
                'horario' => $horario,
            ]);
        } catch (\Illuminate\Database\Eloquent\ModelNotFoundException $e) {
            return response()->json([
                'exito' => false,
                'mensaje' => 'Horario no encontrado',
            ], 404);
        } catch (\InvalidArgumentException $e) {
            return response()->json([
                'exito' => false,
                'mensaje' => $e->getMessage(),
            ], 400);
        }
    }

    /**
     * PATCH /api/horarios/{id}/ocupacion
     * Replica: @PatchMapping("/{id}/ocupacion") → actualizarOcupacion()
     */
    public function actualizarOcupacion(Request $request, int $id): JsonResponse
    {
        $validated = $request->validate([
            'ocupado' => 'required|boolean',
        ]);

        try {
            $horario = $this->horarioService->actualizarOcupacion($id, $validated['ocupado']);
            return response()->json([
                'exito' => true,
                'mensaje' => 'Estado de ocupación actualizado',
                'horario' => $horario,
            ]);
        } catch (\Illuminate\Database\Eloquent\ModelNotFoundException $e) {
            return response()->json([
                'exito' => false,
                'mensaje' => 'Horario no encontrado',
            ], 404);
        }
    }

    /**
     * DELETE /api/horarios/{id}
     * Replica: @DeleteMapping("/{id}") → eliminarHorario()
     */
    public function destroy(int $id): JsonResponse
    {
        try {
            $this->horarioService->eliminarHorario($id);
            return response()->json([
                'exito' => true,
                'mensaje' => 'Horario eliminado correctamente',
            ]);
        } catch (\InvalidArgumentException $e) {
            return response()->json([
                'exito' => false,
                'mensaje' => $e->getMessage(),
            ], 404);
        }
    }
}

File: integraupt-web/Integraupt-Backend/BackendHorarios/src/main/java/com/horarios/controller/HorarioController.java
package com.horarios.controller;

import com.horarios.dto.ActualizarOcupacionRequest;
import com.horarios.dto.HorarioRequest;
import com.horarios.dto.HorarioResponse;
import com.horarios.dto.HorarioSemanalResponse;
import com.horarios.service.HorarioService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.bind.annotation.PatchMapping;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/horarios")
@CrossOrigin(origins = "*")
public class HorarioController {

    private final HorarioService horarioService;

    public HorarioController(HorarioService horarioService) {
        this.horarioService = horarioService;
    }

    @GetMapping
    public List<HorarioResponse> listarHorarios() {
        return horarioService.listarTodos();
    }

    @GetMapping("/{id}")
    public ResponseEntity<Map<String, Object>> obtenerHorario(@PathVariable Integer id) {
        return horarioService.buscarPorId(id)
                .map(horario -> ResponseEntity.ok(respuestaExitosa("Horario encontrado", "horario", horario)))
                .orElse(ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(respuestaError("Horario no encontrado")));
    }

    @GetMapping("/espacio/{espacioId}")
    public List<HorarioResponse> listarPorEspacio(@PathVariable Integer espacioId) {
        return horarioService.listarPorEspacio(espacioId);
    }

    @GetMapping("/espacio/{espacioId}/semanal")
    public List<HorarioSemanalResponse> obtenerHorarioSemanal(@PathVariable Integer espacioId) {
        return horarioService.obtenerHorarioSemanalPorEspacio(espacioId);
    }

    @GetMapping("/dia/{diaSemana}")
    public List<HorarioResponse> listarPorDia(@PathVariable String diaSemana) {
        return horarioService.listarPorDia(diaSemana);
    }

    @GetMapping("/disponibles")
    public List<HorarioResponse> listarDisponibles() {
        return horarioService.listarPorOcupacion(false);
    }

    @GetMapping("/ocupados")
    public List<HorarioResponse> listarOcupados() {
        return horarioService.listarPorOcupacion(true);
    }

    @PostMapping
    public ResponseEntity<Map<String, Object>> crearHorario(@Valid @RequestBody HorarioRequest request) {
        HorarioResponse horario = horarioService.crearHorario(request);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(respuestaExitosa("Horario creado correctamente", "horario", horario));
    }

    @PutMapping("/{id}")
    public ResponseEntity<Map<String, Object>> actualizarHorario(@PathVariable Integer id,
                                                                 @Valid @RequestBody HorarioRequest request) {
        try {
            HorarioResponse horario = horarioService.actualizarHorario(id, request);
            return ResponseEntity.ok(respuestaExitosa("Horario actualizado correctamente", "horario", horario));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(respuestaError(ex.getMessage()));
        }
    }

    @PatchMapping("/{id}/ocupacion")
    public ResponseEntity<Map<String, Object>> actualizarOcupacion(@PathVariable Integer id,
                                                                   @Valid @RequestBody ActualizarOcupacionRequest request) {
        try {
            HorarioResponse horario = horarioService.actualizarOcupacion(id, request);
            return ResponseEntity.ok(respuestaExitosa("Estado de ocupación actualizado", "horario", horario));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(respuestaError(ex.getMessage()));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Map<String, Object>> eliminarHorario(@PathVariable Integer id) {
        try {
            horarioService.eliminarHorario(id);
            return ResponseEntity.ok(respuestaExitosa("Horario eliminado correctamente", null, null));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(respuestaError(ex.getMessage()));
        }
    }

    private Map<String, Object> respuestaExitosa(String mensaje, String claveDato, Object dato) {
        Map<String, Object> respuesta = new HashMap<>();
        respuesta.put("exito", true);
        respuesta.put("mensaje", mensaje);
        if (claveDato != null && dato != null) {
            respuesta.put(claveDato, dato);
        }
        return respuesta;
    }

    private Map<String, Object> respuestaError(String mensaje) {
        Map<String, Object> respuesta = new HashMap<>();
        respuesta.put("exito", false);
        respuesta.put("mensaje", mensaje);
        return respuesta;
    }
}